ocaml/typing/jkind.ml

```diff
@@ -261,6 +274,238 @@ end
 
 type sort = Sort.t
 
+(* A *layout* of a type describes the way values of that type are stored at
+   runtime, including details like width, register convention, calling
+   convention, etc. A layout may be *representable* or *unrepresentable*.  The
+   middle/back ends are unable to cope with values of types with an
+   unrepresentable layout. The only unrepresentable layout is `any`, which is
+   the top of the layout lattice. *)
+module Layout = struct
+  module Const = struct
+    type t =
+      | Sort of Sort.const
+      | Any
+
+    let equal c1 c2 =
+      match c1, c2 with
+      | Sort s1, Sort s2 -> Sort.equal_const s1 s2
+      | Any, Any -> true
+      | (Any | Sort _), _ -> false
+
+    let sub c1 c2 : Sub_result.t =
+      match c1, c2 with
+      | _ when equal c1 c2 -> Equal
+      | _, Any -> Sub
+      | Any, Sort _ | Sort _, Sort _ -> Not_sub
+  end
+
+  type t =
+    | Sort of Sort.t
+    | Any
+
+  let max = Any
+
+  let equate_or_equal ~allow_mutation t1 t2 =
+    match t1, t2 with
+    | Sort s1, Sort s2 -> (
+      match Sort.equate_tracking_mutation s1 s2 with
+      | (Equal_mutated_first | Equal_mutated_second) when not allow_mutation ->
+        Misc.fatal_errorf "Jkind.equal: Performed unexpected mutation"
+      | Unequal -> false
+      | Equal_no_mutation | Equal_mutated_first | Equal_mutated_second -> true)
+    | Any, Any -> true
+    | (Any | Sort _), _ -> false
+
+  let sub t1 t2 : Sub_result.t =
+    match t1, t2 with
+    | Any, Any -> Equal
+    | _, Any -> Sub
+    | Any, _ -> Not_sub
+    | Sort s1, Sort s2 -> if Sort.equate s1 s2 then Equal else Not_sub
+
+  let intersection t1 t2 =
+    match t1, t2 with
+    | _, Any -> Some t1
+    | Any, _ -> Some t2
+    | Sort s1, Sort s2 -> if Sort.equate s1 s2 then Some t1 else None
+
+  let of_new_sort_var () =
+    let sort = Sort.new_var () in
+    Sort sort, sort
+
+  let of_sort_for_error sort = Sort sort
+
+  let value = Sort Sort.value
+
+  let void = Sort Sort.void
+
+  let float64 = Sort Sort.float64
+
+  module Debug_printers = struct
+    open Format
+
+    let t ppf = function
+      | Any -> fprintf ppf "Any"
+      | Sort s -> fprintf ppf "Sort %a" Sort.Debug_printers.t s
+  end
+end
+
+(* Whether or not a type is external to the garbage collector *)
+module Externality = struct
+  type t =
+    | External (* not managed by the garbage collector *)
+    | External64 (* not managed by the garbage collector on 64-bit systems *)
+    | Internal (* managed by the garbage collector *)
```

Comment:
So, the way I understand this, we use externality `External` for  for unboxed floats; shouldn't we instead have a different externality for types that *must not* be managed by the garbage collector at opposed to types than can or cannot be managed by it?

Reply:
Missed this somehow; sorry.

This story will likely evolve. I think we'll probably never consult the externality of a non-`value` type. So maybe this ends up somehow being folded in with the `Value` sort. I have not experimented with this. First, I'd like to use externality to affect compilation (that is, check the externality of a value to decide whether to compile it like an `immediate`) and then revisit the possibility of reshuffling this.

In more direct answer to your question: if the sort of a type is not `value`, then it must not be managed by the garbage collector. All such things should also be `External`, but arranging for this in general is hard and unnecessary (because I think we consult `External` only for `value`s).

Reply:
> Should it be a static error to construct a kind with a layout other than value that is not always external?

I originally thought it should be, but this is hard to arrange, because we sometimes have to decide externality before we know what sort a type will have. If we had externality variables, then all would be well, but let's not. So I think externality just is meaningless for non-`value`s.

@Ekdohibs is right, also, once we get to product layouts. For a product of a `value` and some other layout, the externality field will still be relevant.

Reply:
I think there's nothing to action here now. Resolving.